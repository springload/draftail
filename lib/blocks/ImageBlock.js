import PropTypes from 'prop-types';
import React, { Component } from 'react';

import { ICONS, ENTITY_TYPE } from '../api/constants';

import Icon from '../components/Icon';

/**
 * Editor block to preview and edit images.
 */
class ImageBlock extends Component {
    constructor(props) {
        super(props);

        this.changeText = this.changeText.bind(this);
        this.changeAlignment = this.changeAlignment.bind(this);
        this.renderMediaOptions = this.renderMediaOptions.bind(this);
    }

    changeText(e) {
        const { onSave } = this.props;

        onSave({
            alt: e.currentTarget.value,
        });
    }

    changeAlignment(e) {
        const { onSave } = this.props;

        onSave({
            alignment: e.currentTarget.value,
        });
    }

    renderMediaOptions() {
        const { entity, entityConfig } = this.props;
        const { alt, alignment } = entity.getData();
        const imageFormats = entityConfig.imageFormats || [];

        return (
            <div className="RichEditor-media-options">
                <div className="RichEditor-grid">
                    <div className="RichEditor-col">
                        <label className="u-block">
                            <h4>Alt text</h4>
                            <p>
                                <input
                                    type="text"
                                    value={alt || ''}
                                    onChange={this.changeText}
                                />
                            </p>
                        </label>
                    </div>
                    <div className="RichEditor-col">
                        <h4>Image alignment</h4>
                        <p>
                            {imageFormats.map(format => {
                                return (
                                    <label key={format.value}>
                                        <input
                                            type="radio"
                                            value={format.value}
                                            onChange={this.changeAlignment}
                                            checked={format.value === alignment}
                                        />
                                        {format.label}
                                    </label>
                                );
                            })}
                        </p>
                    </div>
                </div>
            </div>
        );
    }

    render() {
        const { entity, isActive, onClick } = this.props;
        const { src, alt } = entity.getData();

        /* eslint-disable springload/jsx-a11y/no-static-element-interactions */
        return (
            <div>
                <Icon
                    icon={ICONS[ENTITY_TYPE.IMAGE]}
                    className="RichEditor-media-icon"
                />

                <div onClick={onClick} className="RichEditor-media-container">
                    <span className="RichEditor-media-preview">
                        <img src={src} alt={alt || ''} />
                    </span>
                </div>

                {isActive ? this.renderMediaOptions() : null}
            </div>
        );
    }
}

ImageBlock.propTypes = {
    entityConfig: PropTypes.object.isRequired,
    entity: PropTypes.object.isRequired,
    isActive: PropTypes.bool.isRequired,
    onClick: PropTypes.func.isRequired,
    onSave: PropTypes.func.isRequired,
};

export default ImageBlock;
