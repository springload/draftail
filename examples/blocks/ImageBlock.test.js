import React from 'react';
import { shallow } from 'enzyme';
import ImageBlock from '../blocks/ImageBlock';

const mockProps = {
    entity: {
        getData() {
            return {
                src: 'http://www.example.com/example.png',
                alt: 'Test alt',
                alignment: 'right',
            };
        },
    },
    entityConfig: {},
    isActive: false,
    onClick: jest.fn(),
    onSave: jest.fn(),
};

describe('ImageBlock', () => {
    it('empty', () => {
        expect(
            shallow(
                <ImageBlock
                    {...mockProps}
                    entity={{
                        getData() {
                            return {};
                        },
                    }}
                />,
            ),
        ).toMatchSnapshot();
    });

    it('basic', () => {
        expect(shallow(<ImageBlock {...mockProps} />)).toMatchSnapshot();
    });

    it('#isActive', () => {
        expect(
            shallow(<ImageBlock {...mockProps} isActive={true} />),
        ).toMatchSnapshot();
    });

    it('empty #isActive', () => {
        expect(
            shallow(
                <ImageBlock
                    {...mockProps}
                    isActive={true}
                    entity={{
                        getData() {
                            return {};
                        },
                    }}
                />,
            ),
        ).toMatchSnapshot();
    });

    it('#entityConfig.imageFormats', () => {
        expect(
            shallow(
                <ImageBlock
                    {...mockProps}
                    entityConfig={{
                        imageFormats: [
                            { label: 'Left', value: 'left' },
                            { label: 'Right', value: 'right' },
                        ],
                    }}
                    isActive={true}
                />,
            ),
        ).toMatchSnapshot();
    });

    it('#alignment', () => {
        expect(
            shallow(
                <ImageBlock
                    {...mockProps}
                    entityConfig={{
                        imageFormats: [
                            { label: 'Left', value: 'left' },
                            { label: 'Right', value: 'right' },
                        ],
                    }}
                    isActive={true}
                />,
            ).setState({ alignment: 'left' }),
        ).toMatchSnapshot();
    });

    it('changeText', () => {
        const wrapper = shallow(<ImageBlock {...mockProps} isActive={true} />);

        wrapper.find('[type="text"]').simulate('change', {
            currentTarget: {
                value: 'new alt',
            },
        });

        expect(mockProps.onSave).toHaveBeenCalledWith(
            expect.objectContaining({ alt: 'new alt' }),
        );
    });

    it('changeAlignment', () => {
        const wrapper = shallow(
            <ImageBlock
                {...mockProps}
                isActive={true}
                entityConfig={{
                    imageFormats: [
                        {
                            value: 'left',
                            label: 'Left',
                        },
                        { value: 'center', label: 'Center' },
                    ],
                }}
            />,
        );

        wrapper
            .find('[type="radio"]')
            .at(1)
            .simulate('change', {
                currentTarget: {
                    value: 'center',
                },
            });

        expect(mockProps.onSave).toHaveBeenCalledWith(
            expect.objectContaining({ alignment: 'center' }),
        );
    });
});
