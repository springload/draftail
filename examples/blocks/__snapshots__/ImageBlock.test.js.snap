// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageBlock alt 1`] = `
<MediaBlock
  alt=""
  block={Object {}}
  blockProps={
    Object {
      "editorState": Object {},
      "entity": Object {
        "getData": [Function],
      },
      "entityConfig": Object {},
      "onChange": [Function],
    }
  }
  src="example.png"
>
  <label
    className="ImageBlock__field"
  >
    <p>
      Alt text
    </p>
    <input
      onChange={[Function]}
      type="text"
      value="Test"
    />
  </label>
  <button
    className="Tooltip__button"
  >
    Edit
  </button>
  <button
    className="Tooltip__button"
  >
    Remove
  </button>
</MediaBlock>
`;

exports[`ImageBlock renders 1`] = `
<MediaBlock
  alt=""
  block={Object {}}
  blockProps={
    Object {
      "editorState": Object {},
      "entity": Object {
        "getData": [Function],
      },
      "entityConfig": Object {},
      "onChange": [Function],
    }
  }
  src="example.png"
>
  <label
    className="ImageBlock__field"
  >
    <p>
      Alt text
    </p>
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </label>
  <button
    className="Tooltip__button"
  >
    Edit
  </button>
  <button
    className="Tooltip__button"
  >
    Remove
  </button>
</MediaBlock>
`;
